File: app/src/main/java/com/bearcave/passageplanning/base/database/withcustomkey/CRUDWithCustomKey.java
package com.bearcave.passageplanning.base.database.withcustomkey;

import java.util.List;

public interface CRUDWithCustomKey<T, Key> {
    /**
     * @param element
     * @return the row ID of the newly inserted row
     */
    long insert(T element);
    
    T read(Key id);
    List<T> readAll();


    /**
     * @param element
     * @return the number of rows affected
     */
    int update(T element);


    /**
     * @param element
     * @return the number of rows affected if a whereClause is passed in, 0 otherwise.
     */
    int delete(T element);
}

File: app/src/main/java/com/bearcave/passageplanning/base/database/withcustomkey/CRUDWithCustomKey.kt
package com.bearcave.passageplanning.base.database.withcustomkey

interface CRUDWithCustomKey<T, Key> {
    /**
     * @param element
     * *
     * @return the row ID of the newly inserted row
     */
    fun insert(element: T): Long

    fun read(id: Key): T
    fun readAll(): List<T>


    /**
     * @param element
     * *
     * @return the number of rows affected
     */
    fun update(element: T): Int


    /**
     * @param element
     * *
     * @return the number of rows affected if a whereClause is passed in, 0 otherwise.
     */
    fun delete(element: T): Int
}
